namer/rancher/src/main/scala/io/buoyant/namer/rancher/RancherClient.scala

```diff
@@ -0,0 +1,140 @@
+package io.buoyant.namer.rancher
+
+import com.fasterxml.jackson.databind.{ObjectMapper, PropertyNamingStrategy}
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+
+import com.twitter.finagle.{Service, Http, Address, Stack}
+import com.twitter.finagle.http
+import com.twitter.finagle.service.Retries
+import com.twitter.finagle.tracing.NullTracer
+import com.twitter.finagle.stats.{Stat, StatsReceiver}
+
+import com.twitter.logging.Logger
+import com.twitter.util.Activity.State
+import com.twitter.util.{Timer, Var, Activity, Duration, Future, Closable, Return, Throw}
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+case class PortMapping(
+  val ip: String,
+  val publicPort: Int,
+  val privatePort: Int,
+  val protocol: String
+) {
+  // This is a bit of a hack - and should have more error-checking. But it
+  // makes the JSON parsing work.
+  // If anyone knows how to do a better constructor - or JSON Deserializer,
+  // that would be nice
```

Comment:
I'll look into doing a custom deserialiser first, but if that doesn't work out I'll definitely look into improving the constructor this way.

I had a lot of issues trying to overload the constructor of a case-class so I could use a regex, so to make progress I added the very ugly kludge to make some progress. My recollection was that the constructor for case-classes had to start with a call to `this(...)`, but I might be mistaken.

Reply:
I think that for a `case class`, rather than providing a constructor, you probably want a [companion object](https://twitter.github.io/scala_school/basics2.html#object) with an `apply()` method.